test/helpers/foundry/extract.ts

```diff
@@ -0,0 +1,211 @@
+import { ok } from 'node:assert/strict';
+import { createHash } from 'node:crypto';
+import { createWriteStream } from 'node:fs';
+import { rename, mkdir, rm } from 'node:fs/promises';
+import { Agent as HttpAgent } from 'node:http';
+import { Agent as HttpsAgent } from 'node:https';
+import { join, basename, extname, relative } from 'node:path';
+import { pipeline } from 'node:stream/promises';
+import { Minipass } from 'minipass';
+import { extract as extractTar } from 'tar';
+import { Open, type Source, type Entry } from 'unzipper';
+import { say, noop } from './utils';
+import { startDownload } from './download';
+import { Extension, Binary } from './types';
+
+/**
+ * Extracts the binaries from the given URL and writes them to the destination.
+ *
+ * @param url - The URL of the archive to extract the binaries from
+ * @param binaries - The list of binaries to extract
+ * @param dir - The destination directory
+ * @param checksums - The checksums to verify the binaries against
+ * @returns The list of binaries extracted
+ */
+
+export async function extractFrom(
+  url: URL,
+  binaries: string[],
+  dir: string,
+  checksums: { algorithm: string; binaries: Record<Binary, string> } | null,
+) {
+  const extract = url.pathname.toLowerCase().endsWith(Extension.Tar)
+    ? extractFromTar
+    : extractFromZip;
+  // write all files to a temporary directory first, then rename to the final
+  // destination to avoid accidental partial extraction. We don't use
+  // `os.tmpdir` for this because `rename` will fail if the directories are on
+  // different file systems.
+  const tempDir = `${dir}.downloading`;
+  const rmOpts = { recursive: true, maxRetries: 3, force: true };
+  try {
+    // clean up any previous in-progress downloads
+    await rm(tempDir, rmOpts);
+    // make the temporary directory to extract the binaries to
+    await mkdir(tempDir, { recursive: true });
+    const downloads = await extract(
+      url,
+      binaries,
+      tempDir,
+      checksums?.algorithm,
+    );
+    ok(downloads.length === binaries.length, 'Failed to extract all binaries');
+
+    const paths: string[] = [];
+    for (const { path, binary, checksum } of downloads) {
+      if (checksums) {
+        say(`verifying checksum for ${binary}`);
+        const expected = checksums.binaries[binary as Binary];
```

Comment:
This fails on Windows because `binary = 'anvil.exe'` and the data is in `checksums.binaries['anvil']`, not `checksums.binaries['anvil.exe']`

Reply:
![image](https://github.com/user-attachments/assets/dc61fd3f-24bc-4434-9f54-b87c3ce8650b)